File: apps/api/src/routes/documents.ts
import { Hono } from 'hono'
import { prisma } from '../lib/prisma.js'
import { storageService } from '../services/storage.service.js'
import { ingestionService } from '../services/ingestion.service.js'
import { createDocumentSchema } from '@agentbuddy/shared'
import { sanitizeFileName } from '../lib/sanitize.js'
import { validateBody } from '../lib/validate.js'

const app = new Hono()

// All documents
app.get('/documents', async (c) => {
  const documents = await prisma.document.findMany({
    include: { workspace: { select: { name: true } } },
    orderBy: { createdAt: 'desc' },
  })
  return c.json({ success: true, data: documents })
})

app.get('/workspaces/:workspaceId/documents', async (c) => {
  const { workspaceId } = c.req.param()
  const folderIdParam = c.req.query('folderId')

  const where: { workspaceId: string; folderId?: string | null } = { workspaceId }
  if (folderIdParam !== undefined) {
    where.folderId = folderIdParam === 'null' ? null : folderIdParam
  }

  const documents = await prisma.document.findMany({
    where,
    orderBy: { createdAt: 'desc' },
  })
  return c.json({ success: true, data: documents })
})

app.post('/workspaces/:workspaceId/documents', async (c) => {
  const { workspaceId } = c.req.param()
  const contentType = c.req.header('content-type') ?? ''

  if (contentType.includes('multipart/form-data')) {
    const formData = await c.req.formData()
    const file = formData.get('file') as File | null
    const folderId = (formData.get('folderId') as string | null) || null

    if (!file) {
      return c.json({ success: false, error: 'No file provided' }, 400)
    }

    const ext = file.name.split('.').pop()?.toUpperCase() ?? 'TXT'
    const typeMap: Record<string, string> = {
      PDF: 'PDF',
      DOCX: 'DOCX',
      MD: 'MARKDOWN',
      TXT: 'TXT',
      HTML: 'HTML',
    }
    const docType = typeMap[ext] ?? 'TXT'

    const key = `documents/${workspaceId}/${Date.now()}-${sanitizeFileName(file.name)}`
    const buffer = Buffer.from(await file.arrayBuffer())
    await storageService.upload(key, buffer, file.type)

    const document = await prisma.document.create({
      data: {
        title: file.name,
        workspaceId,
        type: docType,
        status: 'PENDING',
        fileUrl: key,
        folderId,
      },
    })

    await ingestionService.enqueue(document.id, key)

    return c.json({ success: true, data: document }, 201)
  }

  // JSON body — create document with inline content
  const body = await c.req.json()
  const data = validateBody(createDocumentSchema, body)
  const document = await prisma.document.create({
    data: {
      title: data.title,
      workspaceId,
      type: data.type ?? 'TXT',
      status: 'READY',
      content: data.content ?? null,
      folderId: data.folderId ?? null,
    },
  })

  return c.json({ success: true, data: document }, 201)
})

app.get('/workspaces/:workspaceId/documents/:docId', async (c) => {
  const { docId } = c.req.param()
  const document = await prisma.document.findUnique({ where: { id: docId } })
  if (!document) {
    return c.json({ success: false, error: 'Document not found' }, 404)
  }
  return c.json({ success: true, data: document })
})

app.patch('/workspaces/:workspaceId/documents/:docId', async (c) => {
  const { docId } = c.req.param()
  const body = await c.req.json()
  const document = await prisma.document.update({
    where: { id: docId },
    data: { folderId: body.folderId ?? null },
  })
  return c.json({ success: true, data: document })
})

app.post('/workspaces/:workspaceId/documents/:docId/reingest', async (c) => {
  const { docId } = c.req.param()
  const document = await prisma.document.findUnique({ where: { id: docId } })
  if (!document) {
    return c.json({ success: false, error: 'Document not found' }, 404)
  }

  // Delete existing chunks so they get recreated
  await prisma.documentChunk.deleteMany({ where: { documentId: docId } })

  // Reset status
  await prisma.document.update({
    where: { id: docId },
    data: { status: 'PENDING', processingStep: null, processingPct: 0, chunkCount: 0 },
  })

  await ingestionService.enqueue(docId, document.fileUrl ?? undefined)

  return c.json({ success: true })
})

app.delete('/workspaces/:workspaceId/documents/:docId', async (c) => {
  const { docId } = c.req.param()
  await prisma.document.delete({ where: { id: docId } })
  return c.json({ success: true, data: { id: docId } })
})

export default app

File: backend/src/clawbuddy/routers/documents.py
"""Document routes.

Replaces: apps/api/src/routes/documents.ts
"""

from __future__ import annotations

import time
from typing import Any

from fastapi import APIRouter, Depends, Request, UploadFile, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from clawbuddy.db.models import Document, DocumentChunk, Workspace
from clawbuddy.db.session import get_db
from clawbuddy.lib.responses import fail, ok
from clawbuddy.lib.sanitize import sanitize_file_name
from clawbuddy.schemas.document import CreateDocumentInput
from clawbuddy.services.storage import storage_service

router = APIRouter(tags=["Documents"])

_EXT_TYPE_MAP: dict[str, str] = {
    "PDF": "PDF",
    "DOCX": "DOCX",
    "MD": "MARKDOWN",
    "TXT": "TXT",
    "HTML": "HTML",
}


@router.get("/documents")
async def list_all_documents(db: AsyncSession = Depends(get_db)) -> dict[str, Any]:
    result = await db.execute(
        select(Document)
        .options(selectinload(Document.workspace))
        .order_by(Document.created_at.desc())
    )
    documents = result.scalars().all()
    return ok(documents)


@router.get("/workspaces/{workspace_id}/documents")
async def list_workspace_documents(
    workspace_id: str,
    folderId: str | None = None,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    stmt = select(Document).where(Document.workspace_id == workspace_id)
    if folderId is not None:
        folder_val = None if folderId == "null" else folderId
        stmt = stmt.where(Document.folder_id == folder_val)
    stmt = stmt.order_by(Document.created_at.desc())
    result = await db.execute(stmt)
    return ok(list(result.scalars().all()))


@router.post("/workspaces/{workspace_id}/documents", status_code=status.HTTP_201_CREATED)
async def create_document(
    workspace_id: str,
    request: Request,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    content_type = request.headers.get("content-type", "")

    if "multipart/form-data" in content_type:
        form = await request.form()
        file: UploadFile | None = form.get("file")  # type: ignore[assignment]
        folder_id = form.get("folderId") or None

        if not file or not file.filename:
            return fail("No file provided", status_code=400)

        ext = file.filename.rsplit(".", 1)[-1].upper() if "." in file.filename else "TXT"
        doc_type = _EXT_TYPE_MAP.get(ext, "TXT")

        key = f"documents/{workspace_id}/{int(time.time() * 1000)}-{sanitize_file_name(file.filename)}"
        contents = await file.read()
        await storage_service.upload(key, contents, file.content_type or "application/octet-stream")

        doc = Document(
            title=file.filename,
            workspace_id=workspace_id,
            type=doc_type,
            status="PENDING",
            file_url=key,
            folder_id=folder_id,
        )
        db.add(doc)
        await db.commit()
        await db.refresh(doc)

        # Enqueue ingestion
        from clawbuddy.services.ingestion import ingestion_service

        await ingestion_service.enqueue(doc.id, key)

        return ok(doc)

    # JSON body — inline content
    body = await request.json()
    data = CreateDocumentInput.model_validate(body)
    doc = Document(
        title=data.title,
        workspace_id=workspace_id,
        type=data.type.value,
        status="READY",
        content=data.content,
        folder_id=data.folder_id,
    )
    db.add(doc)
    await db.commit()
    await db.refresh(doc)
    return ok(doc)


@router.get("/workspaces/{workspace_id}/documents/{doc_id}")
async def get_document(
    workspace_id: str,
    doc_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    result = await db.execute(select(Document).where(Document.id == doc_id))
    doc = result.scalar_one_or_none()
    if not doc:
        return fail("Document not found", status_code=404)
    return ok(doc)


@router.patch("/workspaces/{workspace_id}/documents/{doc_id}")
async def update_document(
    workspace_id: str,
    doc_id: str,
    body: dict[str, Any],
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    result = await db.execute(select(Document).where(Document.id == doc_id))
    doc = result.scalar_one()
    doc.folder_id = body.get("folderId")
    await db.commit()
    await db.refresh(doc)
    return ok(doc)


@router.post("/workspaces/{workspace_id}/documents/{doc_id}/reingest")
async def reingest_document(
    workspace_id: str,
    doc_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    result = await db.execute(select(Document).where(Document.id == doc_id))
    doc = result.scalar_one_or_none()
    if not doc:
        return fail("Document not found", status_code=404)

    # Delete existing chunks
    await db.execute(
        select(DocumentChunk).where(DocumentChunk.document_id == doc_id)
    )
    from sqlalchemy import delete

    await db.execute(delete(DocumentChunk).where(DocumentChunk.document_id == doc_id))

    # Reset status
    doc.status = "PENDING"
    doc.processing_step = None
    doc.processing_pct = 0
    doc.chunk_count = 0
    await db.commit()

    from clawbuddy.services.ingestion import ingestion_service

    await ingestion_service.enqueue(doc_id, doc.file_url)

    return ok(None)


@router.delete("/workspaces/{workspace_id}/documents/{doc_id}")
async def delete_document(
    workspace_id: str,
    doc_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    result = await db.execute(select(Document).where(Document.id == doc_id))
    doc = result.scalar_one()
    await db.delete(doc)
    await db.commit()
    return ok({"id": doc_id})
